src/Shared/PropertyHelper/PropertyHelper.cs

```diff
@@ -492,36 +473,63 @@ private static PropertyHelper CreateInstance(PropertyInfo property)
         }
 
         result = filteredProperties.ToArray();
-        visiblePropertiesCache.TryAdd(type, result);
+        visiblePropertiesCache?.TryAdd(type, result);
         return result;
     }
 
-    protected static PropertyHelper[] GetProperties(
-        [DynamicallyAccessedMembers(DynamicallyAccessedMemberTypes.PublicProperties | DynamicallyAccessedMemberTypes.NonPublicProperties)] Type type,
-        Func<PropertyInfo, PropertyHelper> createPropertyHelper,
-        ConcurrentDictionary<Type, PropertyHelper[]> cache)
+    /// <summary>
+    /// Creates and caches fast property helpers that expose getters for every public get property on the
+    /// specified type.
+    /// </summary>
+    /// <param name="type">The type to extract property accessors for.</param>
+    /// <param name="cache">The cache to store results in. Use <see cref="PropertiesCache"/> to use the default cache. Use <see langword="null"/> to avoid caching.</param>
+    /// <returns>A cached array of all public properties of the specified type.
+    /// </returns>
+    // There isn't a way to represent trimmability requirements since for type since we unwrap nullable types.
+    [RequiresUnreferencedCode("This API is not trim safe.")]
+    public static PropertyHelper[] GetProperties(
+        Type type,
+        ConcurrentDictionary<Type, PropertyHelper[]>? cache)
     {
         // Unwrap nullable types. This means Nullable<T>.Value and Nullable<T>.HasValue will not be
         // part of the sequence of properties returned by this method.
         type = Nullable.GetUnderlyingType(type) ?? type;
```

Comment:
Apparently this makes the whole code path not trimmer friendly. The DynamicallyAccessedMembers annotations are on `type`, but our attempt to unwrap nullable types messes with it because the returned value does not have any annotations. I wonder if I'm holding it wrong or if the trimmer needs to special case this API.

@vitek-karas / @eerhardt?

Reply:
This is a tricky one. From the trimmer (IL) perspective `Nullable<TestType>` is a completely different type from `TestType` and calling `GetProperties` on one has no implications to the other. So applying `PublicProperties` annotation to the `Nullable<TestType>` doesn't mean that it will preserve properties on `TestType`.

In order to make this work the way this code would need we would have to:
* Special case `Nullable<T>` and when applying annotations to it, we would have to apply them to both `Nullable<>` and `T`.
* Special case `Nullable<>.GetUnderlyingType` and propagate the annotation from the input `type` to the return value of this method.

It's definitely doable.

Technically this is a bigger problem - one can imagine that similar logic could be applied to for example `Lazy<T>`, or the Roslyn's `Optional<T>` and so on. But I don't know if it's worth introducing a new kind of annotation to be able to apply this to any generic (something like `class Lazy<[PropagateAnnotationToThis] T>`).

I created https://github.com/dotnet/linker/issues/2662 to track this.
Please let us know how important is this - meanwhile I'll initiate a discussion on the linker team to come up with the best solution to the problem.